### ISSUE TITLE
[BUG] pbench-fio returns 1 - No count for get_mean_hash. at BenchPostprocess.pm

### ISSUE THREAD

**Issue by ldoktor** [distributed-system-analysis/pbench#2953]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Running fio jobs on multiple clients sometimes fails to produce the "results.json" file as the postprocess of one iteration fails with:

    fio job complete
    No count for get_mean_hash. at /opt/pbench-agent/bench-scripts/postprocess/BenchPostprocess.pm line 150.

### To Reproduce

benchmark_bin=/usr/local/bin/fio pbench-fio  --block-sizes=4,1024 --iodepth=1 --numjobs=1 --ramptime=10 --runtime=60 --samples=5 --targets=/fio --test-types=read,write --clients=192.168.122.136,192.168.122.101,192.168.122.107,192.168.122.173

### Actual Results

Usually it works well but about 1/5 attempts one of the sample postprocess fails causing the job postprocess fail as well and return 1.

### Expected Results

pbench-fio should finish without issues.

### Additional information

Running the https://download.copr.fedorainfracloud.org/results/portante/pbench/ pbench version on RHEL8.

**Comment by ldoktor** [comment]
I haven't seen any error messages on the clients, also attaching one failed fio-result.txt [fio-result.txt](https://github.com/distributed-system-analysis/pbench/files/9179174/fio-result.txt)

**Comment by webbnh** [comment]
Hi @ldoktor,

Sorry for the trouble, and thanks for the report!

You said it fails only occasionally?  If you do the same run multiple times, where one works and one fails, could you get us a successful result that corresponds to the one that fails?  (Or, if that's hard, could you attach another, successful result?)

The problem you're seeing results when the postprocessing code tries to take the mean of a series of datapoints which happens to be empty...so, in order to determine why it's failing, we have to figure out what is _not there_.  That will be easier to pick out if we also have a run where it **_is_** there to compare it to.

There are several series, and I'm not sure how to figure out which one is missing; but, once we have that, then we might be able to figure out _why_ it's missing sometimes.

**Comment by portante** [comment]
@ldoktor can you try using the officially released version of the Pbench Agent v0.71 from https://copr.fedorainfracloud.org/coprs/ndokos/pbench/?  See the [release notes](https://github.com/distributed-system-analysis/pbench/releases/tag/v0.71.0) for the reference to that COPR repository in the "Installation" section.  Let me know how that behaves.

**Comment by ldoktor** [comment]
Hello Peter, I was wrong, the installed pbench version was pbench-agent-0.71.0-3g85910732a.noarch from ndokos:pbench, so that should be the officially released version.

**Comment by portante** [comment]
> Hello Peter, I was wrong, the installed pbench version was pbench-agent-0.71.0-3g85910732a.noarch from ndokos:pbench, so that should be the officially released version (updated the description).

Great, thanks @ldoktor.  I'll dig in on this a bit.

**Comment by portante** [comment]
@ldoktor how much skew exists between the clocks on all clients involved?

It looks like the test ran for about 50 seconds and we have a clock skew of over a minute.

**Comment by ldoktor** [comment]
Hello @portante, all guests are created just before we execute the (first) test so it shouldn't be great. But the skew could be caused by hung tasks (from time to time I see issues like:

```
[ 2335.711096] INFO: task auditd:746 blocked for more than 120 seconds.
[ 2335.712248]       Not tainted 4.18.0-372.9.1.el8.x86_64 #1
[ 2335.713098] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 2335.714211] task:auditd          state:D stack:    0 pid:  746 ppid:     1 flags:0x00004320
[ 2335.715342] Call Trace:
[ 2335.715729]  __schedule+0x2d1/0x830
[ 2335.811281]  schedule+0x35/0xa0
[ 2335.811777]  __xfs_log_force_lsn+0x19e/0x1d0 [xfs]
[ 2335.812577]  ? wake_up_q+0x70/0x70
[ 2335.813091]  ? xfs_file_fsync+0x1ac/0x200 [xfs]
[ 2335.813811]  xfs_log_force_lsn+0x8f/0x120 [xfs]
[ 2335.814487]  xfs_file_fsync+0x1ac/0x200 [xfs]
[ 2335.815139]  do_fsync+0x38/0x70
[ 2336.189776]  __x64_sys_fsync+0x10/0x20
[ 2336.190370]  do_syscall_64+0x5b/0x1a0
[ 2336.202774]  entry_SYSCALL_64_after_hwframe+0x65/0xca
[ 2336.219178] RIP: 0033:0x7f522b15d017
[ 2336.219932] Code: Unable to access opcode bytes at RIP 0x7f522b15cfed.
[ 2336.220939] RSP: 002b:00007f52288abd00 EFLAGS: 00000293 ORIG_RAX: 000000000000004a
[ 2336.222117] RAX: ffffffffffffffda RBX: 0000000000000005 RCX: 00007f522b15d017
[ 2336.223294] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000005
[ 2336.224440] RBP: 0000000000000000 R08: 0000000000000000 R09: 00007f52288ac700
[ 2336.225572] R10: 0000000000000000 R11: 0000000000000293 R12: 00007ffc7e944a1e
[ 2336.226656] R13: 00007ffc7e944a1f R14: 0000000000000000 R15: 00007f52288abe80
```

Still I don't see fio complaining about it, or is it? Have you identified what metric is causing the failure?

**Comment by ldoktor** [comment]
Thank you, @webbnh, I know a bit more about how pbench works with the data now (still digesting it) and have some pointers on how to improve my side (what to collect as well as how to tweak my setup). Now what about pbench, do you think it can be improved somehow to better handle such issues? I can see 2 places for improvement:

1) descriptive message about no-overlap in the postprocessed data
2) automated re-run or iteration skip in such case (similarly to uperf which re-runs jittery iterations)

What do you think? (in the mean-time I do plan on adding time-stamps + time sync and re-try stuff; note all this mess is due to known issues with host-page-cache where 16G of host memory is not enough on intense-reading)

**Comment by webbnh** [comment]
> what about pbench, do you think it can be improved somehow to better handle such issues

Absolutely!  The present behavior is an ungraceful mode of failure:  the existing diagnostic message is not actionable for the end-user, and the hard-"crash" and loss of processing of all subsequent samples is inappropriate.

The failure should result in a diagnostic which is meaningful to the user (e.g., pointing out that there will be no aggregate data, due to the lack of concurrency between the clients' runs), and the post-processing should complete for the current sample (albeit producing no aggregate data) and should continue with processing the next sample.

Once you are able to determine whether you can ameliorate the problem by better sync'ing the clients' clocks, we can consider adding advice on how to do that to the diagnostic message.

**Comment by ldoktor** [comment]
Thank you, @webbnh, for all the help. I played with this issue with some extra debug messages, the times are synchronized properly, but the host-page-cache blocks one (or multiple) guests for more than a minute meaning the fio job get's executed after the other jobs already finished.

So, this is basically our bug (or perhaps setup-issue), but as you mentioned it'd be nice to detect that and either re-run, skip or abort the execution without running the consequent variants. Do you want me to close this issue so you can create an issue just for the post-processing or can this issue serve for that?

**Comment by webbnh** [comment]
@ldoktor, we'll keep this issue open for now, as it points out a couple of things in Pbench which should be addressed.  (For scheduling purposes, we'll track this in PBENCH-843.)
